### ISSUE TITLE
State object filtering doesn't work for property value is None

### ISSUE THREAD

**Issue by thaije** [matrx-software/matrx#216]
**Describe the bug**
If I search for an object with a property, it is not possible to filter for objects which have None for that property: 
E.g. if I have the following two objects: 
```
{object_id="obj1", "specific_property": "foofoo"}
{object_id="obj2", "specific_property": None}
{object_id="obj3"}
```

See the examples below:
```
state[{'specific_property': None}] # returns [obj1 & obj2]
state[{'specific_property': 'foofoo'}] # returns obj1
state[{'specific_property': 'lala'}] # returns no objects
```

Thus, it is not possible to find the object with property `specific_property` and value `None`. 

**To Reproduce**
See above

**Expected behavior**
Returns all objects with the `specific_property` with as value `None`

**Comment by jwaa** [comment]
@thaije Good point. I don't think it will be a difficult fix, but you never know... Is it a bug that stands in the way of something (e.g. does it need to be fixed as soon as possible)? If so, do you think you can fix it or do you want me to take a look?

**Comment by thaije** [comment]
@jwaa Nice speedy reaction 😄 I need it for MHC to look for free beds, such as `beds = state[{'name': "Bed_top", 'room': self.medical_care, 'assigned_patient': None}]`. However, that doesn't work as described above. For now I can do a workaround by looking for `beds = state[{'name': "Bed_top", 'room': self.medical_care, 'assigned_patient': 'free'}]`. 

Is the above something you programmed yourself? Or is it some standard processing of the code you used which we need to change? If it's the latter and it takes me more time to find where to fix it, it might be more efficient for you to fix it. Otherwise I can also do it :)

**Comment by jwaa** [comment]
@thaije But ofcourse 👍 

Just went quickly through the code and can't find the reason for this bug, in fact it seems that this should work (every `None` value gets translate to "just look if it has this property, no matter the value"). In other words, it will probably take a bit of effort to find the bug. 

Does it work if you just do `beds = state[{'assigned_patient': None}]`? This way I know whether it actually depends on the interpretation of `None` or if the bug is in the combination of multiple conditions.

In the mean time, setting a different value for empty beds seems to be a good work around.

**Comment by thaije** [comment]
@jwaa `beds = state[{'assigned_patient': None}]` also gives back all beds, including this with assigned_patient is not None.

However, if it works as you describe here it is logical that it doesn't work right? 'every None value gets translate to "just look if it has this property, no matter the value"'
It should be '"look if it has this property **and** this value', or did I misunderstand?

**Comment by jwaa** [comment]
@thaije You're correct! I am the one who misunderstood, I thought the query didn't return anything but instead it returns everything with that property. 

Then we found the bug/ However, it is a tough bug to work around due to the current choice on how to represent a query to the state (a dictionary of property names with their values, and if only a property name is given the value is set to `None`). An easy workaround is defining a different value that represents "just a property name, with no value given" but that just relocates the bug to a different value.

I will think about a more structured solution to this, in the mean time it means that setting a property value in your query to `None` means "find any object that has this property, even if the value of that property is something else then `None`". With the advised work around; don't set property values to None.
